### ISSUE TITLE
A database client should be based on asyncio

### ISSUE THREAD

**Issue by elrik75** [ClickHouse/clickhouse-connect#141]
Major DBs have their asyncio client nowdays:
- postgresql: asyncpg
- redis: aioredis
- kafka: aiokafka
- etc

It's logical for any I/O access for python to be async.
Note that there is an unmaintained async client for clickhouse: https://github.com/maximdanilchenko/aiochclient
Having an official async client for clickhouse with cython and full-featured will be great !

**Comment by genzgd** [comment]
Agreed, this should have had its own enhancement request/issue before now.

**Comment by pkit** [comment]
@elrik75 No, it shouldn't yet.
`asyncio` is so brain-dead right now that it would need a couple of years and a lot of changes to become usable.

**Comment by vsbaldeev** [comment]
Hello everyone! 
As I understand official aio client for python is not developed yet.
So, which programming language and library should I use to make asynchronous requests to clickhouse ?

**Comment by genzgd** [comment]
The `await/async` keywords are part of the Python standard `asyncio` library and work fine with `clickhouse-connect`.  The main caveat is that ClickHouse doesn't allow concurrent queries in the same session, so you should disable the `autogenerate_session_id` common setting before calling `get_client`.  See the explanation here:  https://clickhouse.com/docs/en/integrations/python#managing-clickhouse-session-ids

**Comment by genzgd** [comment]
To clarify, I see this issue as an enhancement request to either create a clean async wrapper around the library that feels more "async native" or even better, use a true async HTTP client.  The urllib3 library doesn't use any async io so in that sense `clickhouse-connect` doesn't really do things in a real event driven way.  But you can still use `clickhouse-connect` with async code.

**Comment by araa47** [comment]
+1 , feel like this is a required feature for a modern python library

**Comment by chrisprobst** [comment]
> @elrik75 No, it shouldn't yet. `asyncio` is so brain-dead right now that it would need a couple of years and a lot of changes to become usable.

Mind elaborating what you mean with brain-dead? It's very usable and used everywhere. What aspect of it is brain-dead and better than thread-blocking urllib3 (which means much more brain-dead to me)?

**Comment by zerafachris** [comment]
Hi @elrik75 @araa47 @genzgd @vsbaldeev , 
there seems to be a package which wraps on the old clickhouse driver here https://github.com/long2ice/asynch . 
Maybe this could help?

**Comment by pkit** [comment]
> Mind elaborating what you mean with brain-dead? It's very usable and used everywhere. What aspect of it is brain-dead and better than thread-blocking urllib3 (which means much more brain-dead to me)?

See here: https://stackoverflow.com/questions/46827007/runtimeerror-this-event-loop-is-already-running-in-python

Essentially brain-dead is the "all or nothing" implementation. No project can be converted to async ever without a complete rewrite of that project and all of the libraries it uses. And the libraries that use these libraries.
No nested tasks are allowed, no gradual I/O conversion, full rewrite only.

**Comment by elrik75** [comment]
> Essentially brain-dead is the "all or nothing" implementation. 

That's correct. The point is more: what kind of DB client should CH support then? sync or async? What kind makes more sense? My point of view (because I work on backend servers) is that a sync IO lib is not very helpful.

**Comment by genzgd** [comment]
@alexted  We in the process of defining the roadmap for `clickhouse-connect` and other integrations, but for the moment this work is not planned for the immediate future.  The current thinking is that true async support would involve swapping out to the `httpx` HTTP client with integrated async support (instead of the venerable urllib3), but that is of course a fairly major change.

To be clear, this is definitely work we want to do, but it may take a while to come to the top of the priority list given all of the work on other projects.  As always, community contributions to help us get moving in the right direction are always appreciated. :)

**Comment by samuelcolvin** [comment]
In case anyone is struggling with this, if the postgres connection port is open, you can connect to clickhouse with psycopg3:

```py
    conn = await psycopg.AsyncConnection.connect(
        dbname='default',
        user='default',
        password='...',
        host='localhost',
        port='9005',
        cursor_factory=psycopg.AsyncClientCursor,
        autocommit=True,
    )
```

You need `cursor_factory=psycopg.AsyncClientCursor, autocommit=True` to get psycopg to use the simple postgres protocol since clickhouse doesn't support the extended protocol, for the same reason you can't use asyncpg.

**Comment by genzgd** [comment]
I've added an [example](https://github.com/ClickHouse/clickhouse-connect/blob/main/examples/run_async.py) of how to run `clickhouse-connect` queries asynchronously (it includes a semaphore just for fun as well).  Based on very limited testing, this solution still significantly outperforms other ClickHouse "async" libraries.  You're still running on a single Python thread at a time, but the HTTP requests are obviously very much I/O bound and the Cython transform code is extremely fast, so the GIL isn't as big a problem as you might think.

If you do experiment with similar solutions, please report your results (good or bad) here.

**Comment by slvrtrn** [comment]
Since [0.7.16](https://github.com/ClickHouse/clickhouse-connect/releases/tag/v0.7.16), ClickHouse-Connect provides a convenience `AsyncClient` wrapper over the standard `Client`, so that it is no longer required to write your own. See [this new entry in the docs](https://clickhouse.com/docs/en/integrations/python#asyncclient-wrapper). This[ async usage example](https://github.com/ClickHouse/clickhouse-connect/blob/v0.7.16/examples/run_async.py) was also updated.

**Comment by samuelcolvin** [comment]
> Since [0.7.16](https://github.com/ClickHouse/clickhouse-connect/releases/tag/v0.7.16), ClickHouse-Connect provides a convenience `AsyncClient` wrapper over the standard `Client`, so that it is no longer required to write your own. See [this new entry in the docs](https://clickhouse.com/docs/en/integrations/python#asyncclient-wrapper). The [async usage example](https://github.com/ClickHouse/clickhouse-connect/blob/v0.7.16/examples/run_async.py) was also updated.

This is very unfortunate. As I think I said to @tbragin, pretending to have any async client by just running the sync in a thread pool is an own goal for ClickHouse. It'll often be slower than just using the sync client. You'd be better off either recommending `loop.run_in_executor` or the HTTP API and httpx.

https://github.com/ClickHouse/clickhouse-connect/blob/ee3af795dcbfdc372facd944cf2212afb962ef16/clickhouse_connect/driver/asyncclient.py#L79-L88

To actually solve this properly, you need to either:
* implement a real async client from the ground up
* or (preferable) support [Arrow Flight SQL](https://arrow.apache.org/docs/format/FlightSql.html) in ClickHouse, and then rely on a real high quality Arrow Flight SQL client library in Python, there's [already](https://arrow.apache.org/rust/arrow_flight/sql/index.html) a Rust client for Arrow Flight SQL, so building a Python one should be pretty easy

**Comment by mshustov** [comment]
@samuelcolvin, As you can see, the issue is not closed. We will support asyncio natively, however the implementation will take time. `AsyncClient` wrapper is created to allow using the client today

**Comment by genzgd** [comment]
@samuelcolvin Putting aside the Arrow Flight question for the moment, I'm not sure what problem you are trying to "solve".  Queries run using the wrapper yield the asyncio event loop when waiting for the HTTP I/O.  So your main thread will spend its time in the CPU bound parts of the query (generally, transforming data).  The fact that `urllib3` itself doesn't have an async API doesn't change how the asyncio event loop behaves while waiting for network data.

I can assure you that using `clickhouse-connect` with the async wrapper will outperform any code using a client  based on `httpx` and a text format like JSON or CSV.  So if you are getting an `async` API that yields the main event loop thread while waiting for I/O, and has high performance, what exactly is missing?

**Comment by genzgd** [comment]
> This is a very poor answer.

Because?  Do you have an actual response to "what is missing"?  It seems like you are hung up on some "theoretically correct" vision regardless of the performance or usability of the library.

**Comment by pkit** [comment]
@samuelcolvin unfortunately httpx is not ready for prime time https://github.com/encode/httpx/issues/3215
Needs some work. Why do you suggest it?

_(Ironically the main httpx problem of the current implementation lies in what I mentioned before: total unreadiness of python itself for a proper async performance/usability, they opted for using anyio just to support different backends)_

**Comment by thewhaleking** [comment]
> Basically a `ThreadPoolExecutor` solution will work fine for simple use cases when nothing goes wrong, but WILL become annoying for end developers in production applications. It's not a coincidence that companies trying hard to have a great DX like OpenAI, Anthropic and Google are adopted httpx, rather than wrapping sync code in a thread pool.

Can confirm we are currently experiencing this pain. Clickhouse has been pretty good overall, but I think we're going to need to switch to Postgres as we scale because the asyncio client is just atrocious for anything at scale.
